experimental/rsocket-src/RSocketServer.cpp

```diff
@@ -0,0 +1,113 @@
+  }
+  void resumeSocket(
+      std::shared_ptr<FrameTransport> frameTransport,
+      ResumeIdentificationToken,
+      ResumePosition,
+      Executor& executor) {
+    //      onSetup_(std::move(frameTransport), std::move(setupPayload));
+  }
+
+ private:
+  OnSetupNewSocket onSetup_;
+};
+
+RSocketServer::RSocketServer(
+    std::unique_ptr<ConnectionAcceptor> connectionAcceptor)
+    : lazyAcceptor(std::move(connectionAcceptor)) {}
+
+void RSocketServer::start(OnAccept onAccept) {
+  if (!acceptor_) {
+    LOG(INFO) << "RSocketServer => initialize connection acceptor on start";
+    acceptor_ = std::make_unique<RSocketConnection>([
+      this,
+      onAccept = std::move(onAccept)
+    ](std::shared_ptr<FrameTransport> frameTransport,
+      ConnectionSetupPayload setupPayload,
+      Executor & executor_) {
+      LOG(INFO) << "RSocketServer => received new setup payload";
+
+      std::shared_ptr<RequestHandler> requestHandler;
+      try {
+        requestHandler = onAccept(
+            std::make_unique<ConnectionSetupRequest>(std::move(setupPayload)));
+      } catch (RSocketError& e) {
+        // TODO emit ERROR ... but how do I do that here?
+        frameTransport->close(
+            folly::exception_wrapper{std::current_exception(), e});
+        return;
+      }
+      LOG(INFO) << "RSocketServer => received request handler";
+
+      auto rs = StandardReactiveSocket::disconnectedServer(
+          // we know this callback is on a specific EventBase
+          executor_,
+          std::move(requestHandler),
+          Stats::noop());
+
+      rs->clientConnect(std::move(frameTransport));
+
+      // register callback for removal when the socket closes
+      rs->onClosed([ this, rs = rs.get() ](const folly::exception_wrapper& ex) {
+        removeSocket(*rs);
+        LOG(INFO) << "RSocketServer => removed closed connection";
+      });
+
+      // store this ReactiveSocket
+      addSocket(std::move(rs));
+    });
+  } else {
+    throw std::runtime_error("RSocketServer.start already called.");
+  }
+  lazyAcceptor->start([ this, onAccept = std::move(onAccept) ](
+      std::unique_ptr<DuplexConnection> duplexConnection, Executor & executor) {
+    LOG(INFO) << "RSocketServer => received new connection";
+
+    LOG(INFO) << "RSocketServer => going to accept duplex connection";
+    // the callbacks above are wired up, now accept the connection
+    acceptor_->acceptConnection(std::move(duplexConnection), executor);
+  });
+}
+
+void RSocketServer::startAndPark(OnAccept onAccept) {
+  start(std::forward<OnAccept>(onAccept));
+  // now block this thread
+  std::unique_lock<std::mutex> lk(m);
+  // if shutdown gets implemented this would then be released by it
+  cv.wait(lk, [] { return false; });
+}
+
+void RSocketServer::removeSocket(ReactiveSocket& socket) {
+  LOG(INFO) << "RSocketServer => remove socket from vector";
+  reactiveSockets_.erase(std::remove_if(
```

Comment:
I attempted not storing these references, but then it is really awkward for the user. 

This code works:

```cpp
auto rs = rsf->connect().get();
    rs->requestStream(Payload("Jane"), s);
    s->awaitTerminalEvent();
```

but this one doesn't using future.then:

```cpp
rsf->connect().then([s](std::shared_ptr<RSocketRequester> rs) {
      rs->requestStream(Payload("Bob"), s);
    });
    s->awaitTerminalEvent();
```

because as soon as the 'then' scope exits, it destroys everything.

Due to the asynchrony of this, what do you recommend as the solution here?